From: Victor Seva <vseva@sipwise.com>
Date: Thu, 19 Jan 2017 17:13:12 +0100
Subject: [PATCH] TT#9317 mod_mam: don't store bodyless chat messages

* implement store hints
  https://prosody.im/issues/issue/750
  mod_mam doesn't honor forced storage hint.
  When a message without a body is received mod_mam always ignores the forced
  storage hint, even if <store xmlns='urn:xmpp:hints'/> is present.
  This can cause issue with OMEMO, which sends message elements without a body

Change-Id: I590b1fb9bd95afdce6a117778052d11d8102f718
---
 plugins/mod_mam/mod_mam.lua | 23 +++++++++++++++++------
 1 file changed, 17 insertions(+), 6 deletions(-)

diff --git a/plugins/mod_mam/mod_mam.lua b/plugins/mod_mam/mod_mam.lua
index faff0dd..759a5e2 100644
--- a/plugins/mod_mam/mod_mam.lua
+++ b/plugins/mod_mam/mod_mam.lua
@@ -285,19 +285,30 @@ local function message_handler(event, c2s)
 		event.stanza = stanza;
 	end
 
-	-- We store chat messages or normal messages that have a body
-	if not(orig_type == "chat" or (orig_type == "normal" and stanza:get_child("body")) ) then
-		log("debug", "Not archiving stanza: %s (type)", stanza:top_tag());
-		return;
-	end
-
 	-- or if hints suggest we shouldn't
 	if not stanza:get_child("store", "urn:xmpp:hints") then -- No hint telling us we should store
+		-- We store chat messages or normal messages that have a body
+		local body = stanza:get_child("body");
+		if not(orig_type == "chat" or (orig_type == "normal" and body)) then
+			log("debug", "Not archiving stanza: %s (type)", stanza:top_tag());
+			return;
+		elseif (orig_type == 'chat' and not body) then
+			log("debug", "Not archiving stanza: %s (type), has no body",
+				stanza:top_tag());
+			return;
+		end
+
 		if stanza:get_child("no-permanent-store", "urn:xmpp:hints")
 			or stanza:get_child("no-store", "urn:xmpp:hints") then -- Hint telling us we should NOT store
 			log("debug", "Not archiving stanza: %s (hint)", stanza:top_tag());
 			return;
 		end
+	else
+		log("debug", "store hint detected");
+		if orig_type == 'error' then
+			log("debug", "Not archiving stanza: %s (type)", stanza:top_tag());
+			return;
+		end
 	end
 
 	-- Check with the users preferences
